Import from old repository

pavolloffay opened the pull request:
Import from old repository https://github.com/pavolloffay/opentracing-java-integrations

pavolloffay commented:
@bensigelman @yurishkuro @adriancole could you please review? 

To all packages applies: 
```
 <groupId>io.opentracing.contrib</groupId>
 <artifactId>opentracing-jaxrs-XXX</artifactId>
```

bensigelman commented:
Three^W Four things up front:

1. Thank you!
1. I am having a crazy week and cannot review this in the next 36h
1. ... but I have starred this in my inbox and will def get to it
1. Thank you!

yurishkuro commented:
@vprithvi could you please review and compare with jaeger-jaxrs?

If we could, I would love to remove explicit jaeger-jaxrs and just depend on opentracing-contrib, but I expect there are some differences where it's not straightforward, so let's open an issue in our module and list them out and figure out how to solve.

vprithvi commented on jaxrs/src/main/java/io/opentracing/contrib/jaxrs/server/SpanServerRequestFilter.java:
```diff
@@ -0,0 +1,78 @@
+
+            Tracer.SpanBuilder spanBuilder =
+                    tracer.buildSpan(operationName.orElse(
+                            URLUtils.path(requestContext.getUriInfo().getRequestUri()).get()));
```
`URLUtils#path` uses `URI#getPath` to determine the path of the request, which could lead to an explosion in the number of operation names because `#getPath` returns the path of the real request, and not that of the `@Path` annotation. 

Consider the request method below, 
```java
@PUT
@Path("/hello/{worldName}")
Response putWorld(@PathParam("worldName"} name) {
...
}
```
When someone calls `PUT /hello/earth`, `#getPath` would return `/hello/earth`. 

Depending on the tracing UI used, having a high cardinality of operation names could make analysis difficult

vprithvi commented on jaxrs/src/main/java/io/opentracing/contrib/jaxrs/client/SpanClientRequestFilter.java:
```diff
@@ -0,0 +1,76 @@
+        tracer.inject(span.context(), Format.Builtin.TEXT_MAP, new TextMapInjectAdapter(spanContext));
+
+        spanContext.forEach((key, value) ->
+                clientRequestContext.getHeaders().add(key, value));
```
This could be rolled into your `TextMapInjectAdapter` (or something similar) so that it acts directly on the headers

adriancole commented on jaxrs/pom.xml:
```diff
@@ -0,0 +1,38 @@
+    <!-- Logging -->
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
```
logging used here is quite basic, why not use jul and avoid pinning a slf4j version?

adriancole commented on jaxrs/pom.xml:
```diff
@@ -0,0 +1,38 @@
+    <dependency>
+      <groupId>javax.servlet</groupId>
+      <artifactId>javax.servlet-api</artifactId>
+      <version>${version.javax.servlet-api}</version>
```
maybe make this provided? jaxrs2 doesn't need servlet 3. iirc provided stops escalating versions

adriancole commented on jaxrs/src/main/java/io/opentracing/contrib/jaxrs/client/SpanDecorator.java:
```diff
@@ -0,0 +1,34 @@
+/**
+ * @author Pavol Loffay
+ */
+public interface SpanDecorator {
```
I've made a type similar to this. it tends to only need to do tags. something to think about, but constraining to tags might be tighter

adriancole commented:
also meta-note... java-jaxrs I suppose that's mainly to allow github module search filtering on name vs incidental classification. I'd probably name it opentracing-jaxrs2 as when it is checked out, I get that directory name (unlike go world which more often lives in the complete github namespace). In other words java-jaxrs is redundant about java, but says nothing about opentracing

adriancole commented:
>
> @bensigelman <https://github.com/bensigelman> @yurishkuro
> <https://github.com/yurishkuro> @adriancole
> <https://github.com/adriancole> could you please review?
>
for you? sure :P

pavolloffay commented on examples/wildfly-swarm/README.md:
````diff
@@ -0,0 +1,16 @@
+
+## Example Requests
+```shell
+$ curl -ivX GET 'http://localhost:3000/hello'
````
:+1: I will use `MockTracer` with console logging. 

https://github.com/opentracing/opentracing-java/pull/76

pavolloffay commented:
PR updated, added `OperationNameProvider` moved to java 7, JUL, removed swarm and dropwizard example apps. 

`<artifactId>opentracing-jaxrs</artifactId>` renamed to `<artifactId>opentracing-jaxrs2</artifactId>`

bensigelman commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/SpanWrapper.java:
```diff
@@ -0,0 +1,31 @@
+        }
+    }
+
+    public boolean isFinished() {
```
nit: synchronized, right?

bensigelman commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/client/SpanDecorator.java:
```diff
@@ -0,0 +1,36 @@
+
+    void decorateRequest(ClientRequestContext containerRequestContext, Span span);
+
+    void decorateResponse(ClientResponseContext containerRequestContext, Span span);
```
nit: `containerResponseContext`

bensigelman commented on examples/spring-boot/pom.xml:
```diff
@@ -0,0 +1,102 @@
+
+    <!-- OpenTracing impl -->
+    <dependency>
+      <groupId>org.hawkular.apm</groupId>
```
@pavolloffay https://github.com/opentracing/opentracing-java/tree/0.20.4

jpkrohling commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/server/CurrentSpan.java:
```diff
@@ -0,0 +1,22 @@
+    private HttpServletRequest request;
+
+    public Span injectedSpan() {
+        SpanWrapper spanWrapper = (SpanWrapper) request.getAttribute(SpanServerRequestFilter.SPAN_PROP_ID);
```
You have several of those `#getAttribute` casting to other value in your code. Instead of doing unsafe casts all around the code, I'd suggest creating an utility class that would accept the type you want to receive the value as, logging an exception and returning null if the type is not of the expected value. As such, the target application can still be executed, even if there's a configuration issue.

jpkrohling commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/server/OperationNameProvider.java:
```diff
@@ -0,0 +1,56 @@
+                path = path.substring(1);
+            }
+
+            for (Map.Entry<String, List<String>> entry: pathParameters.entrySet()) {
```
Please, leave a javadoc for the expected behavior with examples. For instance, I'd like to know what's the expected behavior for the URL `/foo/bar/123`, which reaches a method annotated with `@Path("/foo/bar/{id : \\d+}")`, as well as the URL `/foo/bar/name`, which reaches a method annotated with `@Path("/foo/bar/{name: \\w+}")`

jpkrohling commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/server/SpanServerResponseFilter.java:
```diff
@@ -0,0 +1,40 @@
+    @Override
+    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) throws IOException {
+
+        SpanWrapper spanWrapper = (SpanWrapper)requestContext.getProperty(SpanServerRequestFilter.SPAN_PROP_ID);
```
Be forgiving here as well.

jpkrohling commented on jaxrs2/src/main/resources/META-INF/beans.xml:
```diff
@@ -0,0 +1,6 @@
+    xsi:schemaLocation="
+        http://xmlns.jcp.org/xml/ns/javaee
+        http://xmlns.jcp.org/xml/ns/javaee/beans_1_1.xsd" bean-discovery-mode="all">
+</beans>
```
This might sound overkill, but please provide a `jaxrs2` + `jaxrs2-cdi` module instead of `jaxrs2` with `beans.xml`. People having problems with CDI will have to unpack, remove the beans.xml and repack your jar, or manually set your packages to `ignore` on their CDI configuration.

pavolloffay replied:
Do we actually need cdi integration? There are not beans defined.

jpkrohling replied:
Then get rid of this file :)

jpkrohling commented on pom.xml:
```diff
@@ -0,0 +1,51 @@
+  <packaging>pom</packaging>
+
+  <properties>
+    <maven.compiler.source>1.7</maven.compiler.source>
```
I'd just set it to Java 8 instead. Java 7 is EOLed for quite some time now.

pavolloffay replied:
This does not work `javacTask: source release 1.8 requires target release 1.8`

pavolloffay replied:
I will stick with java 7 for now.

pavolloffay commented on integration-tests/apache-cxf/src/test/java/io/opentracing/contrib/jaxrs/itest/cxf/ApacheCXFWildcardOperationNameITest.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * @author Pavol Loffay
+ */
+public class ApacheCXFWildcardOperationNameITest extends AbstractWildcardOperationNameTest {
```
This uses different `OperationNameProvider`.

pavolloffay commented on examples/spring-boot/src/main/java/io/opentracing/contrib/jaxrs/example/spring/boot/App.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * @author Pavol Loffay
+ */
+@SpringBootApplication
```
Maybe not, but this makes it easy to run it from IDE.

pavolloffay commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/client/SpanClientResponseFilter.java:
```diff
@@ -0,0 +1,41 @@
+    private List<SpanDecorator> spanDecorators;
+
+    public SpanClientResponseFilter(List<SpanDecorator> spanDecorators) {
+        this.spanDecorators = spanDecorators;
```
used `Collections.unmodifiableList(spanDecorators)`.

jpkrohling replied:
The usual approach is to copy the list when receiving it, and freeze it when sending it. So, your idea about freezing it when receiving the list is not bad, it's just not "usual", specially when you want to manipulate the list, without affecting the original list.

adriancole commented:
>
> @vprithvi <https://github.com/vprithvi> @adriancole
> <https://github.com/adriancole> is that ok with you?
>
personally I tend towards the least of java 6/floor language level of the
library being instrumented.

adriancole commented:
also I consider this your repo not mine.. make a call with sound
reasoning. I usually optimize for highest leverage and that means
lower language level, especially as so few people will actually touch
this code vs use it (one would hope)

pavolloffay commented on integration-tests/apache-cxf/pom.xml:
```diff
@@ -0,0 +1,33 @@
+    <dependency>
+      <groupId>${project.groupId}</groupId>
+      <artifactId>opentracing-jaxrs-itest-common</artifactId>
+      <version>${project.version}</version>
```
Not a special reason. When I look at `${project.version}` it's more obvious to me that it is the same version as the whole project. I will change it if there will be problems.

adriancole replied:
I also highly prefer this approach (use variables). It makes forks nicer, too.

pavolloffay commented:
PR updated, added generics and javadoc. Does anybody want to re-review :octocat: ?

bensigelman commented on README.md:
```diff
@@ -1 +1,51 @@
+// register this in javax.ws.rs.core.Application
+ServerTracingDynamicFeature.Builder
+    .traceAll(yourPreferredTracer)
+    .withOperationNameProvider(ServerOperationNameProvider.HTTP_METHOD_NAME_PROVIDER)
```
This is a judgement call, but I prefer to have all things like this moved into the domain of the `SpanDecorator`... *if* there are clear factorable patterns emerging there, then maybe those patterns get extracted into more specialized things like `ServerOperationNameProvider`, etc.

pavolloffay replied:
Operation name can be modified by `SpanDecorator` so it can be confusing. I can think of this solutions:

1. allow `SpanDecorator` to modify only tags, logs, baggage(?) and keep `OperationNameProvider`.
2. get rid of `OperationNameProvider` and use only decorators which operate on span.

What do you think?
I think that second one is more flexible.

pavolloffay commented:
PR updated. 

 @bensigelman I removed `operationNameProvider` and using only `spanDecorator` for operationName.

@vprithvi @yurishkuro @adriancole @bensigelman could you please re-review and if it's fine merge?

yurishkuro commented on README.md:
```diff
@@ -1 +1,49 @@
+// register this in javax.ws.rs.core.Application
+ServerTracingDynamicFeature.Builder
+    .traceAll(yourPreferredTracer)
+    .withStandardTags()
```
what's the purpose of this? In other words, when would someone NOT want standard tags? It sounds like an edge case, perhaps the standard tags should be the default, and if someone really wants to disable them they can call `withoutStandardTags()`

adriancole replied:
I prefer to expose how to install custom tags vs a flag like this. in other words, you either use defaults or you compose something on top of them. This customization has been requested many times from different people in Brave, and will be available shortly. (customization can be a later hook, btw)

yurishkuro commented on README.md:
```diff
@@ -1 +1,49 @@
+
+Response response = jaxRsClient.target("http://localhost/endpoint")
+    .request()
+    .property(TracingProperties.CHILD_OF, parentSpan) // optional, by default new trace is started
```
This is a deal breaker for me personally. Java as a language has a pretty good story for implicit context propagation via thread locals (e.g. via WIP https://github.com/opentracing-contrib/java-activespan), so requiring the users to remember to add code for tracing to every call site is a step back.

adriancole replied:
the thread local stuff breaks often, so I understand the desire to make explicit propagation possible. I agree that the user shouldn't be aware of this by default (iotw, explicit is an option)

pavolloffay replied:
:+1: I will at it once it is merged/released.

yurishkuro commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/client/ClientSpanDecorator.java:
```diff
@@ -0,0 +1,65 @@
+            String url = URIUtils.url(requestContext.getUri());
+            if (url != null) {
+                Tags.HTTP_URL.set(span, url);
+            }
```
peer.xxx standard tags are very easy to add to client calls, and are very valuable.

pavolloffay replied:
Added:
* `peer.port`
* `peer.hostname`

+ old:
* `span.kind`
* `http.method`
* `http.url`
* `http.status_code`

pavolloffay replied:
https://github.com/opentracing-contrib/java-jaxrs/pull/2/commits/92b123e0864397f06415d0bdb3f1612f194bb4d5

yurishkuro commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/client/ClientSpanDecorator.java:
```diff
@@ -0,0 +1,65 @@
+    ClientSpanDecorator HTTP_PATH_OPERATION_NAME = new ClientSpanDecorator() {
+        @Override
+        public void decorateRequest(ClientRequestContext clientRequestContext, Span span) {
+            span.setOperationName(URIUtils.path(clientRequestContext.getUri()));
```
what happens if URL has entity IDs encoded in the path, e.g. `/api/customer/162317263`? If that ID becomes a part of the span name, then this is a very dangerous decorator to provide off the shelf, as it explodes the cardinality of span names.

adriancole replied:
agreed. this is why a lot of operation names are lame in most instrumentation (like the http verb). Getting a clean templated name is ideal, and has been suggested in many forums last year. ex https://github.com/openzipkin/brave/issues/158#issuecomment-234436074

It is tricky as encoding rules for things like JAX-RS templates are probably more complex than they appear. One way out is to go conservative and then tell people to override it. I agree path is probably not a good default.

pavolloffay replied:
By default it's used HTT method name provider, this is just an optional thing.

adriancole commented on README.md:
```diff
@@ -1 +1,49 @@
+@GET
+@Path("/hello")
+@Traced(operationName = "helloRenamed") // optional, by default operation name is provided by ServerSpanDecorator
+public Response hello(@BeanParam CurrentSpan currentSpan) {
```
CurrentSpan makes sense, and is something similarly needed in other async libraries. Ack that the means to inject it is framework specific, and it must be so because thread locals doesn't work in all cases.

adriancole commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/SpanDecorator.java:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @author Pavol Loffay
+ */
+public interface SpanDecorator<IN, OUT>{
```
this interface will catch most customizations you would need.

pavolloffay replied:
I wanted to make it generic for server and client requestContext.

adriancole commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/client/ClientTracingFeature.java:
```diff
@@ -0,0 +1,106 @@
+    public static class Builder {
+        private Tracer tracer;
+        private Client client;
+        private List<ClientSpanDecorator> spanDecorators = new ArrayList<>();
```
note: since you are exposing a list, basically you can expect people to complain about order and other types of problems related to that. Usually I don't, and instead punt to the user to compose their own decorator lists.

pavolloffay replied:
Builder has two methods

`withEmptyDecorators()` which clears array and `withDecorator(decorator)` which adds decorator.
By default in a constructor it adds two decorators one for operation name and other for standard tags.

It makes sense to provide only `Builder#withDecorators(decorators)` I think it's cleaner :+1:

adriancole commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/internal/SpanWrapper.java:
```diff
@@ -0,0 +1,35 @@
+    public synchronized void finish() {
+        if (!finished.get()) {
+            finished.set(true);
+            span.finish();
```
seems span should do this sort of idempotency

pavolloffay replied:
This [1] is for `MockSpan`, however from the discussion I assume it's better not to assume that `Span#finish()` is idempotent (unless it is defined in spec).

[1]: https://github.com/opentracing/opentracing-java/pull/72

adriancole commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/server/ServerHeadersExtractTextMap.java:
```diff
@@ -0,0 +1,70 @@
+ *
+ * @author Pavol Loffay
+ */
+public class ServerHeadersExtractTextMap implements TextMap {
```
personally I'd use a custom carrier instead of routing through textmap, but understand the temptation

pavolloffay commented:
@bensigelman could you please do a hopefully final review and eventually merge?

bensigelman commented:
@pavolloffay FYI I have not forgotten about this and hope to do it tomorrow! Crazy week, sorry for the delay. It's on my list.

bensigelman commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/server/SpanServerRequestFilter.java:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        if (tracer != null) {
+            SpanContext extractedSpanContext = tracer.extract(Format.Builtin.TEXT_MAP,
```
I would have expected `Format.Builtin.HTTP_HEADERS` for both the inject and extract calls... am I missing something?

pavolloffay replied:
Changed to `HTTP_HEADERS`.

bensigelman commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/server/ServerSpanDecorator.java:
```diff
@@ -0,0 +1,103 @@
+
+    /**
+     * As operation name provides HTTP path. If there are path parameters used in URL then
+     * spans for the same requests would have different operation names, therefore use carefully.
```
I'm almost tempted to not-support this since it's such a common stumbling block for people in practice (i.e., not realizing they're enabling a super-high-cardinality operation name).

Just a thought... I could go either way.

pavolloffay replied:
:+1: more people highlighted this. I'm going to remove it there is a better and safe alternative (`HTTP_WILDCARD_PATH_OPERATION_NAME`)

bensigelman commented on jaxrs2/src/main/java/io/opentracing/contrib/jaxrs/server/ServerSpanDecorator.java:
```diff
@@ -0,0 +1,103 @@
+     * {@link io.opentracing.tag.Tags#HTTP_METHOD}, {@link io.opentracing.tag.Tags#HTTP_URL} and
+     * {@link io.opentracing.tag.Tags#HTTP_STATUS}
+     */
+    ServerSpanDecorator STANDARD_TAGS = new ServerSpanDecorator() {
```
I would prefer to make the SPAN_KIND tag both (1) built-in to this package (i.e., non-negotiable) and (2) provided at `tracer.buildSpan(...)`-time.

Part of the rationale is that we *know* this tag should be applied... the other piece is that some impls actually change behavior around SpanContext ids depending on the presence of the SERVER tag *at startSpan() time*. E.g., some zipkin tracers want to make sure the same span id is used on the client and server, and that logic needs to kick in before the span starts in a naive impl.

pavolloffay replied:
Good catch, changes applied.

pavolloffay commented:
@adriancole @bensigelman @yurishkuro @jpkrohling @vprithvi thanks for reviews!
